presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PushdownSubfields.java

```diff
@@ -556,42 +584,194 @@ private static NestedField nestedField(String name)
+                        functionAndTypeManager.getFunctionMetadata(call.getFunctionHandle()).getName().equals(ELEMENT_AT)) {
+                    Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession, functionAndTypeManager);
+                    if (isPruningSubfiedldsPossible(context)) {
+                        addRequiredSubfields(context, subfield);
+                        return null;
+                    }
+                    if (subfield.isPresent()) {
+                        context.subfields.add(subfield.get());
+                    }
+                    else {
+                        call.getArguments().forEach(argument -> argument.accept(this, context));
+                    }
                     return null;
                 }
+                if (isPushDownSubfieldsFromLambdasEnabled) {
+                    LambdaSubfieldContext lambdaSubfieldContextOriginal = context.getLambdaSubfieldContext().orElse(null);
+                    if (functionDescriptor.isAccessingInputValues()
+                            || (functionDescriptor.isTransformingInput()) && !functionDescriptor.getOutputToInputTransformationFunction().isPresent()) {
+                        context.setLambdaSubfieldContext(null);
+                    }
+                    if (!functionDescriptor.isReturningInputValues()) {
+                        context.setLambdaSubfieldContext(new LambdaSubfieldContext());
+                    }
 
-                // visit subscript expressions only
-                Optional<Subfield> subfield = toSubfield(call, functionResolution, expressionOptimizer, connectorSession);
-                if (subfield.isPresent()) {
-                    context.subfields.add(subfield.get());
+                    transformOuterLambdaSubfieldsIfNeeded(context, functionDescriptor);
+
+                    if (context.getLambdaSubfieldContext().isPresent()) {
+                        Set<Integer> argumentIndicesContainingMapOrArray = functionDescriptor.getArgumentIndicesContainingMapOrArray()
+                                .orElse(IntStream.range(0, call.getArguments().size()).boxed().collect(Collectors.toSet()));
+                        Map<Integer, LambdaSubfieldContext> lambdaSubfieldContextMap = argumentIndicesContainingMapOrArray.stream()
+                                .collect(Collectors.toMap(callArgumentIndex -> callArgumentIndex, unused -> context.getLambdaSubfieldContext().get().copy()));
+
+                        for (LambdaDescriptor lambdaDescriptor : functionDescriptor.getLambdaDescriptors()) {
+                            if (!addLambdaSubfields(call, context, lambdaDescriptor, lambdaSubfieldContextMap)) {
+                                return null;
+                            }
+                        }
+                        if (getFunctionName(call.getFunctionHandle()).equals(CARDINALITY)) {
```

Comment:
Let's consider this example
```
CREATE TABLE my_table( array_data ARRAY( ROW( id BIGINT, data VARCHAR ) ) )
SELECT CARDINALITY( array_data ) FROM my_table
```
If all of what we do with the array of rows is obtaining the length of the array than we can just extract request the smallest subfield  from the row. 

the better way to solve it would be to push the cardinality() expression down to the reader. 
Alternatively, we could make the reader to construct the rows with no subfields, but  for now reader treats empty set of the required subfield as `no filter present`.

Reply:
that's not specific to cardinality, though.  That's anything that does not access input and does not return input values (maybe cardinality is the only one like that, but we don't need a special case for cardinality if we're handling those to things properly).  Also, there should be a way to specify no subfields rather than choosing a random subfield like id.

Reply:
Added *noneSubfield*